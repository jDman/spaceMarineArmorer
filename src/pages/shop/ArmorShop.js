import React, { useCallback, useEffect, useState } from 'react';
import axios from 'axios';

import Input from '../../components/form/input/Input';
import Button from '../../components/button/Button';
import PageTitle from '../../components/page-title/PageTitle';

import classes from './ArmorShop.module.css';

const ArmorShop = (props) => {
  const { token } = props;
  const [fetchedArmors, setFetchedArmors] = useState([]);

  const addToCart = async (event, armor) => {
    await props
      .addToCart(event, armor)
      .then((result) => {
        getArmors();
      })
      .catch((err) => console.log(err));
  };

  const getArmors = useCallback(() => {
    return axios
      .get('http://localhost:4000/shop/armors')
      .then((result) => {
        const armors = result.data.armors.map((armor) => ({
          id: armor._id,
          name: armor.name,
          type: armor.type,
          description: armor.description,
          company: armor.company,
          quality: armor.quality,
          protection: armor.protection,
          shield: armor.shield,
          discount: armor.discount,
          cost: armor.cost,
          stock: armor.stock,
          createdBy: armor.createdBy.userName,
          config: {
            elementType: 'input',
            elementConfig: {
              type: 'number',
              name: 'quantity',
              min: 0,
              max: +armor.stock,
              step: 1,
            },
            value: 0,
            valid: false,
            validation: {
              min: 1,
            },
            touched: false,
          },
        }));

        setFetchedArmors(armors);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);

  const checkValidity = (value, rules) => {
    let isValid = true;

    const trimedValue = +value.trim();

    if (rules.min) {
      isValid = trimedValue > 0;
    }

    return isValid;
  };

  const inputChangedHandler = (newQuantity, armorId) => {
    const armors = [...fetchedArmors];

    const updatedArmorDetail = armors.find((armor) => armorId === armor.id);

    updatedArmorDetail.config.value = newQuantity;
    updatedArmorDetail.config.touched = true;

    updatedArmorDetail.config.valid = checkValidity(
      updatedArmorDetail.config.value,
      updatedArmorDetail.config.validation
    );

    armors.map((armor) => {
      if (armor.id === armorId) {
        return {
          ...updatedArmorDetail,
        };
      }

      return armor;
    });

    setFetchedArmors(armors);
  };

  useEffect(() => {
    axios.defaults.headers.common['Authorization'] = 'Bearer ' + token;
    axios.defaults.headers.common['Content-Type'] = 'application/json';
    getArmors();
  }, [getArmors, token]);

  return (
    <section className={classes.ArmorShop}>
      <PageTitle title="Shop for Armor" />

      <ul className={classes.ArmorShopList}>
        {fetchedArmors.map((armor) => (
          <li key={armor.id} className={classes.ArmorShopListItem}>
            <div className={classes.ArmorShopListItemContainer}>
              <p>
                <strong>Name</strong>: {armor.name}
              </p>
              <p>
                <strong>Type</strong>: {armor.type}
              </p>
              <p>
                <strong>Description</strong>: {armor.description}
              </p>
              <p>
                <strong>Company</strong>:{' '}
                {armor.company.trim().split('_').join(' ')}
              </p>
              <p>
                <strong>Quality</strong>: {armor.quality}
              </p>
              <p>
                <strong>Protection</strong>: {armor.protection}
              </p>
              <p>
                <strong>Shield</strong>: {armor.shield}
              </p>
              <p>
                <strong>Discount</strong>: {armor.discount}
              </p>

              <p>
                <strong>Cost</strong>: {armor.cost}
              </p>

              <small>Created By: {armor.createdBy}</small>
            </div>

            <div className={classes.ArmorShopListItemContainer}>
              <form
                onSubmit={(event) => addToCart(event, armor)}
                className={classes.ArmorShopListForm}
              >
                <Input
                  elementType={armor.config.elementType}
                  elementConfig={armor.config.elementConfig}
                  classes={classes.ArmorShopListFormElement}
                  value={armor.config.value}
                  invalid={!armor.config.valid}
                  touched={armor.config.touched}
                  disabled={armor.stock < 1}
                  hasValidation={armor.config.validation}
                  changed={(event) =>
                    inputChangedHandler(event.target.value, armor.id)
                  }
                />

                <Button
                  type="submit"
                  classes={
                    armor.stock > 0
                      ? classes.ArmorShopAddToCartButton
                      : [
                          classes.ArmorShopAddToCartButton,
                          classes.ArmorShopAddToCartButtonDisabled,
                        ]
                  }
                  disabled={!armor.config.valid}
                  buttonText={armor.stock > 0 ? 'Add to Cart' : 'Out of Stock'}
                />
              </form>
            </div>
          </li>
        ))}
      </ul>
    </section>
  );
};

export default ArmorShop;
