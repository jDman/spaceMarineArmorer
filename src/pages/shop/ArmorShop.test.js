import React from 'react';
import renderer, { act } from 'react-test-renderer';
import axios from 'axios';

import ArmorShop from './ArmorShop';

jest.mock('axios');

describe('<ArmorShop />', () => {
  it('should render ArmorShop page with empty armor array returned from server', async () => {
    const result = {
      data: {
        armors: [],
        totalItems: 0,
      },
    };
    const component = renderer.create(<ArmorShop />);

    await act(async () => {
      await axios.get.mockImplementation(() => Promise.resolve(result));
    });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render ArmorShop page with results returned from server for armor', async () => {
    const result = {
      data: {
        armors: [
          {
            createdBy: {
              userId: '5e70dfb438cee83fd9e004fd',
              userName: 'Freddy',
            },
            stock: 5,
            shield: 50,
            discount: 30,
            _id: '5e941bab2c9afa63aa5cadb3',
            type: 'helmet',
            name: 'test',
            cost: 200,
            protection: 'low',
            quality: 'low',
            description: 'Low tier trash',
            company: 'starscape_systems',
            createdAt: '2020-04-13T07:58:35.164Z',
            updatedAt: '2020-04-13T07:58:35.164Z',
            __v: 0,
          },
        ],
        totalItems: 1,
      },
    };
    const component = renderer.create(<ArmorShop />);

    await act(async () => {
      await axios.get.mockImplementation(() => Promise.resolve(result));
    });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render ArmorShop page with results returned from server for armor with no stock', async () => {
    const result = {
      data: {
        armors: [
          {
            createdBy: {
              userId: '5e70dfb438cee83fd9e004fd',
              userName: 'Freddy',
            },
            stock: 0,
            shield: 50,
            discount: 30,
            _id: '5e941bab2c9afa63aa5cadb3',
            type: 'helmet',
            name: 'test',
            cost: 200,
            protection: 'low',
            quality: 'low',
            description: 'Low tier trash',
            company: 'starscape_systems',
            createdAt: '2020-04-13T07:58:35.164Z',
            updatedAt: '2020-04-13T07:58:35.164Z',
            __v: 0,
          },
        ],
        totalItems: 1,
      },
    };
    const component = renderer.create(<ArmorShop />);

    await act(async () => {
      await axios.get.mockImplementation(() => Promise.resolve(result));
    });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('updates quantity form state on input change', async () => {
    const result = {
      data: {
        armors: [
          {
            createdBy: {
              userId: '5e70dfb438cee83fd9e004fd',
              userName: 'Freddy',
            },
            stock: 5,
            shield: 50,
            discount: 30,
            _id: '5e941bab2c9afa63aa5cadb3',
            type: 'helmet',
            name: 'test',
            cost: 200,
            protection: 'low',
            quality: 'low',
            description: 'Low tier trash',
            company: 'starscape_systems',
            createdAt: '2020-04-13T07:58:35.164Z',
            updatedAt: '2020-04-13T07:58:35.164Z',
            __v: 0,
          },
        ],
        totalItems: 1,
      },
    };

    const component = renderer.create(<ArmorShop />);

    const componentInstance = component.root;

    await act(async () => {
      await axios.get.mockImplementation(() => Promise.resolve(result));
    });

    act(() => {
      componentInstance
        .findByProps({ name: 'quantity' })
        .props.onChange({ target: { value: '5' } });
    });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
