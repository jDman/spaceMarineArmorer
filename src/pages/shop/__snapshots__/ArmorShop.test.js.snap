// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArmorShop /> should render ArmorShop page with empty armor array returned from server 1`] = `
Array [
  <ul />,
  <p>
    Total: 
    0
  </p>,
]
`;

exports[`<ArmorShop /> should render ArmorShop page with results returned from server for armor 1`] = `
Array [
  <ul>
    <li>
      <p>
        Type: 
        helmet
      </p>
      <p>
        Description: 
        Low tier trash
      </p>
      <p>
        Company:
         
        starscape systems
      </p>
      <p>
        Quality: 
        low
      </p>
      <p>
        Protection: 
        low
      </p>
      <p>
        Shield: 
        50
      </p>
      <p>
        Discount: 
        30
      </p>
      <p>
        Cost: 
        200
      </p>
      <small>
        Created By: 
        Freddy
      </small>
      <form
        onSubmit={[Function]}
      >
        <div
          className="Input"
        >
          <label
            className="Label"
          >
            quantity
          </label>
          <input
            className="InputElement"
            max={99}
            min={0}
            name="quantity"
            onChange={[Function]}
            step={1}
            type="number"
            value={0}
          />
        </div>
        <button
          disabled={true}
          type="submit"
        >
          Add to Cart
        </button>
      </form>
    </li>
  </ul>,
  <p>
    Total: 
    1
  </p>,
]
`;

exports[`<ArmorShop /> updates quantity form state on input change 1`] = `
Array [
  <ul>
    <li>
      <p>
        Type: 
        helmet
      </p>
      <p>
        Description: 
        Low tier trash
      </p>
      <p>
        Company:
         
        starscape systems
      </p>
      <p>
        Quality: 
        low
      </p>
      <p>
        Protection: 
        low
      </p>
      <p>
        Shield: 
        50
      </p>
      <p>
        Discount: 
        30
      </p>
      <p>
        Cost: 
        200
      </p>
      <small>
        Created By: 
        Freddy
      </small>
      <form
        onSubmit={[Function]}
      >
        <div
          className="Input"
        >
          <label
            className="Label"
          >
            quantity
          </label>
          <input
            className="InputElement"
            max={99}
            min={0}
            name="quantity"
            onChange={[Function]}
            step={1}
            type="number"
            value="5"
          />
        </div>
        <button
          disabled={false}
          type="submit"
        >
          Add to Cart
        </button>
      </form>
    </li>
  </ul>,
  <p>
    Total: 
    1
  </p>,
]
`;
